doc/functions/fileset.section.md

```diff
@@ -0,0 +1,117 @@
+# File sets {#sec-fileset}
+
+File sets efficiently represent a set of local files. File sets easily be created, combined and modified for complex behavior. All file set functionality is exposed under `lib.fileset`.
+
+The most basic way to create file sets is by passing a path to `lib.singleton`. The resulting file set depends on the path:
+- If the path points to a file, the result is a file set only consisting of that single file
+- If the path points to a directory, all files in that directory will be in the resulting file set
+
+This matches the list of files that would be imported when using string interpolation on the path, e.g. `"${./Makefile}"` or `"${./src}"`.
+
+Another way to create file sets is by combining existing file sets together, e.g. using `lib.fileset.unions`:
+
+```nix
+with lib.fileset;
+unions [
+  (singleton ./Makefile)
+  (singleton ./src)
+]
+```
+
+For convenience, all fileset operations implicitly coerce path arguments to file sets using `lib.singleton`, so the above is the same as
+```nix
+with lib.fileset;
+unions [
+  ./Makefile
+  ./src
+]
+```
+
+Another way to create file sets is by transforming existing file sets, e.g. using `lib.fileset.suffixFileFilter`:
+```
+with lib.fileset;
+union [
+  ./Makefile
+  (suffixFileFilter ["c" "h"] ./src)
+]
+```
+
+A fileset can be used in a derivation build by passing it as a `src` to any `stdenv.mkDerivation`-based builder. It is then necessary to also set `sourceRoot` to the path of the builder entry point.
+```nix
+stdenv.mkDerivation {
+  src =
+    with lib.fileset;
+    unions [
+      ./Makefile
+      (suffixFileFilter ["c" "h"] ./src)
+    ];
+  sourceRoot = ./.;
```

Comment:
Re-using `src` and `sourceRoot` is one of the key design decisions here. It makes a lot of sense and feels decently natural to people that already know about these attributes. `sourceRoot` is required here. I'd still need to update reference docs for this.

Some alternatives:
```nix
mkDerivation {
  src = with lib.fileset;
    toSource ./. (union [
      ./Makefile
      ./src
    ];
}
```

Problems:
- Not very clear that `./.` is just the entry point
- It would have to change the semantics of what it means to assign a string(-coercible) value to `src`. Previously it would always just copy the directory that string points to and make it writable. But with this it can copy not just the directory but also some parents of it, which is somewhat unintuitive. Not sure if that's a big problem though.

```nix
mkDerivation {
  entryPoint = ./.;
  fileset = with lib.fileset;
    union [
      ./Makefile
      ./src
    ];
}
```

Not a big fan of this, `src` is a clear indicator that we're specifying the source, whereas it's not as clear here.

```nix
mkDerivation {
  src.entryPoint = ./.;
  src.fileset = with lib.fileset;
    union [
      ./Makefile
      ./src
    ];
}
```

Not bad.